modules/apps/static/portal-configuration/portal-configuration-metatype-api/src/test/java/com/liferay/portal/configuration/metatype/bnd/util/ConfigurableUtilTest.java

```diff
@@ -84,33 +86,43 @@ public void testClassInitializationFailure() throws Exception {
 		}
 	}
 
-	@AdviseWith(adviceClasses = ConfigurableUtilAdvice.class)
-	@NewEnv(type = NewEnv.Type.CLASSLOADER)
 	@Test
 	public void testConcurrentCreateConfigurable() throws Exception {
-		Callable<TestConfiguration> callable =
-			() -> ConfigurableUtil.createConfigurable(
-				TestConfiguration.class,
-				Collections.singletonMap(
-					"testReqiredString", "testReqiredString"));
+		CountDownLatch blockingCountDownLatch = new CountDownLatch(1);
 
-		FutureTask<TestConfiguration> futureTask1 = new FutureTask<>(callable);
-		FutureTask<TestConfiguration> futureTask2 = new FutureTask<>(callable);
+		CountDownLatch waitingCountDownLatch = new CountDownLatch(2);
 
-		Thread thread1 = new Thread(
-			futureTask1, "Configurable Util Test Thread 1");
-		Thread thread2 = new Thread(
-			futureTask2, "Configurable Util Test Thread 2");
+		TestClassLoader testClassLoader = new TestClassLoader(
```

Comment:
@dantewang Why @NewEnv(type = NewEnv.Type.CLASSLOADER) not work with your case?

Reply:
@tinatian `ConfigurableUtil` invokes the define class method on the class loader of the configuration class -- in this test, the class loader that loads `TestConfiguration`. Since we replace the define class method in the util with a method (`testDefineClass`) from our `TestClassLoader`, we must ensure that `TestConfiguration` is loaded by this `TestClassLoader`. This is a rule enforced by Java reflection: if the instance class that the method is invoked on does not match the method's declaring class, an error is thrown.

We can't achieve this by directly using new env class loader rule.

By the way, we also tried to replace the `class` field in `Method`, but this causes JVM crash when running the tests on our side.

Reply:
@dantewang You do not need to have the TestClassLoader, what you need is just a method that invokes the original method, you can pass in the original method as a parameter, rather than having your own defineClass method.

Tina.

Reply:
@tinatian I doubt if it is possible, as I believe that we've tried something similar.

The ConfigurableUtil does this on the method in keeps:

```
method.invoke(classLoader, xxxx, xxxx, ....)
```

where the first parameter must be of the class that declares the method. This is enforced by Java reflection.

If we just write a wrapper/delegating method that wraps the real defineClass method, we can't pass this declaring class verification.